### DISCUSSION TITLE
🚨 Action Required: Ensure that you no longer use gcr.io/kubebuilder images

### DISCUSSION THREAD

**Discussion by camilamacedo86** [kubernetes-sigs/kubebuilder#3907] [Announcements]
## ⚠️ **IMPORTANT** ⚠️

**If your project uses gcr.io/kubebuilder/kube-rbac-proxy it will be affected**. Your project may fail to work if the image cannot be pulled. You must move as soon as possible, sometime from early 2025, the GCR will go away.

> Unfortunately, we're unable to provide any guarantees regarding timelines or potential extensions at this time. Images provided under GRC will be unavailable from March 18, 2025, [as per announcement](https://cloud.google.com/artifact-registry/docs/transition/transition-from-gcr). However, `gcr.io/kubebuilder/` may be unavailable before this date due [to efforts to deprecate infrastructure](https://github.com/kubernetes/k8s.io/issues/2647).

## Key Update

[kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy) was historically used to protect the metrics endpoint. However, its usage has been discontinued in [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder).  
The default scaffold now leverages the [`WithAuthenticationAndAuthorization`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/metrics/filters#WithAuthenticationAndAuthorization) feature provided by [Controller-Runtime](https://github.com/kubernetes-sigs/controller-runtime). 

This feature provides integrated support for securing metrics endpoints by embedding authentication (authn) and authorization (authz) mechanisms directly into the controller manager's metrics server, replacing the need for (https://github.com/brancz/kube-rbac-proxy) to secure metrics endpoints.

## Why This Matters

- **Unavailability:** If your project depends on the image, it may no longer work when the image becomes unavailable.  
- **Security:** Unprotected metrics endpoints may expose sensitive data like system performance and application behavior, creating security vulnerabilities.  

## What You Need to Do

Follow some options.

### **Option 1 - Upgrade the Project to the Latest Release (Recommended)**  

Upgrade your project to the latest release by re-scaffolding it and reintegrating your custom code.

- You can use [kubebuilder alpha generate](https://book.kubebuilder.io/reference/rescaffold) to regenerate the project from the `PROJECT` config. 
- This ensures a default implementation (similar to `kube-rbac-proxy`) with options to improve production readiness, such as configuring certificates securely. This approach also allows you to take advantage of other improvements, bug fixes, and the latest updates.

### **Option 2 - Manually Modify Your Project**  

If you prefer not to fully upgrade, modify your project to use the built-in `authn/authz` protection via Controller-Runtime.  

- Follow the instructions in the FAQ: [**"How can I manually change my project to switch to Controller-Runtime's built-in auth protection?"**](https://github.com/kubernetes-sigs/kubebuilder/discussions/3907).

### **Option 3 - Continue Using the `kube-rbac-proxy` Image (Not adopt or promoted by Kubebuilder)**  

If you want to continue using `kube-rbac-proxy`, source the image from an alternative location, **at your own risk**. Examples include:

- [quay.io/brancz/kube-rbac-proxy](https://quay.io/repository/brancz/kube-rbac-proxy)  
- [Red Hat Registry](https://catalog.redhat.com/search?gs&q=kube-rbac-proxy)  **(⚠️ If you are allowed to use it. Ensure that you check its [terms & conditions](https://www.redhat.com/en/about/agreements). [Example](https://catalog.redhat.com/software/containers/rhacm2/kube-rbac-proxy-rhel9/65a6ebe637d382950a5035fc?container-tabs=gti))**

> To mitigate risks, we manually mirror the images to [registry.k8s.io/kubebuilder/kube-rbac-proxy](https://explore.ggcr.dev/?repo=registry.k8s.io%2Fkubebuilder%2Fkube-rbac-proxy). This registry is managed by the #sig-k8s-infra. However, we cannot promote the kube-rbac-proxy images on this registry or recommend their usage since they have been discontinued from the project.

### **Why does the metrics endpoint need to be protected?**

Unprotected metrics endpoints can expose sensitive data, such as system performance and application behavior, to unauthorized users. This can lead to security vulnerabilities where attackers gain insights into the system's operation and exploit weaknesses.

---

### **How can the metrics endpoint be protected?**

The following are some options with details and info for those who were introduced to support and helpers in Kubebuilder.

1. **(Protection enabled by default from release `v4.1.0`)**  
   Use Controller-Runtime's feature [WithAuthenticationAndAuthorization](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/metrics/filters#WithAuthenticationAndAuthorization) to enable `authn/authz` for metrics endpoints. For reference, see the code from the release [4.3.1](https://github.com/kubernetes-sigs/kubebuilder/blob/a9ee3909f7686902879bd666b92deec4718d92c9/testdata/project-v4/cmd/main.go#L89-L152)

2. **(Optional helper introduced from release `v4.2.0`)**  
   Use NetworkPolicies to secure metrics endpoints.  
   Example configuration: [NetworkPolicy Example](https://github.com/kubernetes-sigs/kubebuilder/blob/a9ee3909f7686902879bd666b92deec4718d92c9/testdata/project-v4/config/default/kustomization.yaml#L30-L34)

   Also, feel free to check the ([external code example](https://github.com/prometheus-operator/kube-prometheus/discussions/1907#discussioncomment-3896712)) 

3. **Integrate cert-manager:**  
   Secure the metrics endpoint using TLS encryption with cert-manager. For example, ensure that you use valid certficates such as:
 
   - Reference for the `ServiceMonitor` to integrate it with Prometheus:

   https://github.com/kubernetes-sigs/kubebuilder/blob/e07823e33190b4d3b9339a8273183187b9b50535/testdata/project-v4/config/prometheus/monitor_tls_patch.yaml#L10-L22
  
   Note that we plan add a feature/helper for it in the next release, see the PR: https://github.com/kubernetes-sigs/kubebuilder/pull/4400  
---

### **Why doesn't Network Policy provide the same level of protection as kube-rbac-proxy?**

NetworkPolicy acts as a firewall for pods, controlling traffic flow at the IP or port level. However, it doesn’t handle authentication (`authn`), authorization (`authz`), or encryption like [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy) does.

---

### **How can I manually change my project to switch to Controller-Runtime's built-in auth protection?**

Following these steps should help you resolve the issue. However, there may be a few caveats depending on how old the version used to create the project is, especially if it has never been upgraded to the latest versions by re-creating the project and adding your code changes on top.

1. **Remove kube-rbac-proxy configurations:**
    Ensure that you remove the container with the `name: kube-rbac-proxy`. [Example from an old version](https://github.com/kubernetes-sigs/kubebuilder/blob/3bfc84ec8767fa760d1771ce7a0cb05a9a8f6286/testdata/project-v4/config/default/manager_auth_proxy_patch.yaml#L28-L50) 

2. **Ensure that your manager will have the args for the metrics service and binding at the same port**

- For example, see the metrics service: https://github.com/kubernetes-sigs/kubebuilder/blob/e07823e33190b4d3b9339a8273183187b9b50535/testdata/project-v4/config/default/metrics_service.yaml#L8-L17

- For example, see the arg to be patch to binding the metrics service: https://github.com/kubernetes-sigs/kubebuilder/blob/e07823e33190b4d3b9339a8273183187b9b50535/testdata/project-v4/config/default/manager_metrics_patch.yaml#L2-L4

3. **Enable built-in auth in `main.go`:**

- Update your `main.go` to implement [WithAuthenticationAndAuthorization](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/metrics/filters#WithAuthenticationAndAuthorization). Here’s an example:  

   ```go
   // if the enable-http2 flag is false (the default), http/2 should be disabled
   // due to its vulnerabilities. More specifically, disabling http/2 will
   // prevent from being vulnerable to the HTTP/2 Stream Cancellation and
   // Rapid Reset CVEs. For more information see:
   // - https://github.com/advisories/GHSA-qppj-fm5r-hxr3
   // - https://github.com/advisories/GHSA-4374-p667-p6c8
   disableHTTP2 := func(c *tls.Config) {
	setupLog.Info("disabling http/2")
	c.NextProtos = []string{"http/1.1"}
   }

   if !enableHTTP2 {
	tlsOpts = append(tlsOpts, disableHTTP2)
   }

   metricsServerOptions := metricsserver.Options{
       BindAddress:   metricsAddr,
       SecureServing: secureMetrics,
       TLSOpts:       tlsOpts,
       FilterProvider: filters.WithAuthenticationAndAuthorization,
   }
   ```
   
- For a full implementation example, see the samples under [testdata](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata). For reference, see the code from the release `4.3.1`:
   
https://github.com/kubernetes-sigs/kubebuilder/blob/a9ee3909f7686902879bd666b92deec4718d92c9/testdata/project-v4/cmd/main.go#L89-L152

Note: Please ensure that you disable HTTP/2 by default, as you see in the above example. Disable HTTP/2 still required: https://github.com/kubernetes/kubernetes/issues/121197

5 - **Add e2e tests to validate the metrics endpoint.**
   
Projects created with the latest versions are scaffolded with comprehensive E2E tests, including code to validate the metrics endpoint. [Example](https://github.com/kubernetes-sigs/kubebuilder/tree/v4.3.1/testdata/project-v4/test/e2e). See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/test/e2e/e2e_test.go#L166-L235

Note that you can remove the [Prometheus block](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/test/e2e/e2e_test.go#L180-L183) if you are not providing this integration.

6. **Build and Test Manually:**

Build your project and ensure the metrics endpoint is working and protected with RBAC as expected. The RBAC permissions scaffolded under `config/rbac` should provide the required permissions. However, if you face issues, you might want to look at the last scaffolds to generate them properly. See that under [testdata](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata); we have examples.

**Steps to Verify Metrics with `curl` manually**

1. Create the role-binding

```sh
kubectl create clusterrolebinding <project-name>-metrics-binding \
  --clusterrole=<project-name>-metrics-reader \
  --serviceaccount=<project-name>-system:<project-name>-controller-manager
```

2. Generate the token to authenticate in the metrics

```sh
export TOKEN=$(kubectl create token operator-controller-controller-manager -n olmv1-system)
echo $TOKEN
```

3. Run the curl pod to perform the tests

```sh
kubectl run curl-metrics --rm -it --restart=Never \
  --image=curlimages/curl:7.87.0 -n <project>-system -- /bin/sh
```

4. Call the metrics enpoint using the TOKEN

```sh
curl -v -k -H "Authorization: Bearer $TOKEN" https://<my-project-name>-controller-manager-metrics-service.<my-project-name>-system.svc.cluster.local:8443/metrics
```
   
---

### **❓ Why is this happening?**

The `kube-rbac-proxy` images have been rebuilt and re-tagged by Kubebuilder for an extended period. However, due to infrastructure changes within the Kubernetes ecosystem and the deprecation of Google Cloud Platform’s Container Registry ([details here](https://cloud.google.com/artifact-registry/docs/transition/transition-from-gcr)), continuing to maintain these images is no longer feasible.

Additionally, the project has been in the process of being donated to SIG-Auth for an extended period. Despite these efforts, significant requirements still need to be addressed before SIG-Auth can consider adopting the project. The latest review outlined several tasks being tracked here: [kube-rbac-proxy issue #238](https://github.com/brancz/kube-rbac-proxy/issues/238).

For more details, refer to:  

- [Proposal for this change](https://github.com/kubernetes-sigs/kubebuilder/pull/2345)  
- [Efforts to deprecate infrastructure](https://github.com/kubernetes/k8s.io/issues/2647)  
- [Meta Issue to Discontinue Kube RBAC Proxy in Default Scaffolding](https://github.com/kubernetes-sigs/kubebuilder/issues/3871)  
- [SIG-Auth Pre-Acceptance 2nd Review](https://github.com/brancz/kube-rbac-proxy/issues/238)  

---

### **Why can't we recommend just replacing the kube-rbac-proxy image with the latest version from another trusted source?**

The following are some key reasons. 

- **Deprecated Features**: The [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy)’s self-signed certificate generation is deprecated and unsuitable for production. Future updates will require actual certificates.
- **Security Concerns**: Updating the image alone doesn’t adequately address security needs for protecting metrics endpoints in production environments.
- **Lack of SIG-Auth Support**: The [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy) project is not formally part of SIG-Auth and has been in the process of donation for a long period but is not there yet. For details on what is needed after the second review round, please look at [GitHub Issue #238](https://github.com/brancz/kube-rbac-proxy/issues/238).
- **Better Alternatives Regarding Maintainability**: Controller-runtime provides a better approach in the long-term. Note that when maintainers want to support new k8s versions in their solutions, they upgrade the controller runtime. This means projects using it can quickly adopt new enhancements, bug fixes, and metrics service protection improvements.
- **Kubebuilder No Longer Adopts This Solution**: The usage of [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy) was discontinued mainly for the reasons mentioned above and due [to efforts to deprecate infrastructure](https://github.com/kubernetes/k8s.io/issues/2647). Furthermore, recommending its use or replacing it with a similar image introduces long-term risks, such as reliance on registries that may not remain available or supported indefinitely.

---

**Please update your configurations accordingly to avoid disruptions.** If you have any questions or need further assistance, feel free to ask in this discussion thread or the [Kubebuilder Slack channel](https://communityinviter.com/apps/kubernetes/community#kubebuilder).

For further information, check the metrics section in the documentation: https://book.kubebuilder.io/reference/metrics.

**Comment by camilamacedo86** [comment]
Hi everyone,

We’ve been updating this discussion to provide the latest information, including details about the helpers implemented in Kubebuilder. Additionally, we’ve been enhancing it by, for example, adding more comprehensive steps to help you manually update your project if needed.

🖐️ **Check out the FAQ section: ["How can I manually change my project to switch to Controller-Runtime's built-in auth protection?"](https://github.com/kubernetes-sigs/kubebuilder/discussions/3907) for detailed instructions.**

We will continue to keep this discussion up to date.
If you face any issues, please don’t hesitate to reach out to us.

**Comment by camilamacedo86** [comment]
Hi everyone,

We’ve been updating this discussion to provide the latest information.

**🖐️ Check out the FAQ section: [Why can't we recommend just replacing the kube-rbac-proxy image with the latest version from another trusted source?](https://github.com/kubernetes-sigs/kubebuilder/discussions/3907#discussion-6670554)**

We will continue to keep this discussion up to date.
If you face any issues, please don’t hesitate to reach out to us.

**Comment by camilamacedo86** [comment]
Hey everybody,

For the next release (`4.4.0`, which will support Kubernetes `1.32`), we are introducing a feature that will help users configure the metrics server with certificates, including those provided by cert-manager. 

You might want to take a look at the related PR https://github.com/kubernetes-sigs/kubebuilder/pull/4400 for more details!

Best regards,

**Comment by mattwelke** [comment]
@camilamacedo86 Thanks for the migration guide! We were able to migrate, but we have a question about k8s RBAC if you don't mind clearing things up for us.

We used the approach where you change `main.go` to use [WithAuthenticationAndAuthorization](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/metrics/filters#WithAuthenticationAndAuthorization) from controller-runtime. We have things running where the metrics server appears to be starting correctly. Logs show this:

```
2024-12-05T19:57:49Z    INFO    controller-runtime.metrics      Serving metrics server  {"bindAddress": ":8443", "secure": true}
```

However, we found that we can't make a successful request (it says it's unauthenticated) to the metrics endpoint unless we add a `ClusterRoleBinding` to bind the `metrics-reader` `ClusterRole` to a service account that we have access to that we can use to get a token. We used the controller manager service account to test this. Then, we were able to run a command like `curl -H "Authorization: Bearer $TOKEN" https://10.96.99.213:8443/metrics -k` to see the metrics.

It's worth noting that we don't make use of the metrics from our custom operator yet, we're just following this migration guide.

The example project has a `ClusterRole` (https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/config/rbac/metrics_reader_role.yaml) but not a `ClusterRoleBinding` that binds it to anything.

Is the idea that the example project isn't supposed to be a complete example that shows reading metrics, because everyone will have a different way they need to read metrics? And a different thing reading them, which would have its own service account that would be bound?

**Reply by camilamacedo86** [reply]
Hi @mattwelke,

Thank you for raising this question! I'm glad the migration guide helped you move forward. Let me address your points inline:

> The example project has a ClusterRole ([link](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/config/rbac/metrics_reader_role.yaml)) but not a ClusterRoleBinding that binds it to anything.

Both the `ClusterRole` and `ClusterRoleBinding` are provided in the example. The samples under `testdata` are continuously updated and tested via our CI pipeline. Here's the relevant file you may have missed:

[metrics_auth_role_binding.yaml](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/config/rbac/metrics_auth_role_binding.yaml)

![Screenshot of metrics_auth_role_binding.yaml](https://github.com/user-attachments/assets/ee74aae7-4815-4fa8-bd2a-60562e5ec558)

> Is the idea that the example project isn't supposed to be a complete example that shows reading metrics, because everyone will have a different way they need to read metrics? And a different thing reading them, which would have its own service account that would be bound?

The examples under `testdata` are generated by Kubebuilder's CLI and built as part of our development process. These files represent what the CLI scaffolds for users by default. You can find the commands that generate these samples [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/testdata/generate.sh).

In the latest Kubebuilder releases, the CLI scaffolds a lot to support common use cases, including end-to-end (e2e) testing. For example, the scaffolded e2e tests include checks for the metrics endpoint. You can see this in the [default e2e test file](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/test/e2e/e2e_test.go#L166-L234).

However, how the metrics endpoint is used depends on the specific requirements of each implementation. For instance:

- You might integrate the metrics endpoint with **Prometheus**. (Our default scaffold includes [helpers for this integration](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v4/config/prometheus).)
- You might integrate it with **Grafana**, using the files generated by our [optional plugin for Grafana dashboards](https://kubebuilder.io/plugins/available/grafana-v1-alpha).
- You could use it for other monitoring solutions within your cluster.
- Either you can use it in your e2e tests, such as we use to validate the projects built with it as its features. Refer to how the metrics endpoint is tested in Kubebuilder's CLI itself. Here's an example:  [plugin_cluster_test.go](https://github.com/kubernetes-sigs/kubebuilder/blob/ba2825d47b03d943690d7af586f2175be259f17d/test/e2e/v4/plugin_cluster_test.go#L415-L422)

**To answer your question directly:** The example project doesn't dictate how to consume metrics because this varies across use cases. Instead, the scaffolds provide a solid foundation that you can extend based on your specific needs.

I hope this helps! Let me know if you have any other questions. 😊

**Reply by mattwelke** [reply]
That's super clear. Ty!

**Reply by camilamacedo86** [reply]
Hi @mattwelke 

I also updated the steps to test to try to make it easier.
Thank you a lot for raising that.

**Comment by camilamacedo86** [comment]
Currently, I good solution is also upgrade the operator to the latest **Kubebuilder** release.  
With the help of AI tools, this process might be easier than expected.

**Resources:**
- Kubebuilder AI migration helpers:  
  https://book.kubebuilder.io/migration/ai-helpers.html
- Related discussion in Kubernetes Slack:  
  https://kubernetes.slack.com/archives/CAR30FCJZ/p1772671271796409

After upgrading, maintaining the project becomes much easier if you use the **`alpha update`** command. It helps keep the project up to date without losing your custom changes on top of the generated code.

See:
- https://kubebuilder.io/reference/commands/alpha_update

If that approach doesn’t work for your case, another option is to review how Kubebuilder implements it internally and replicate something similar manually in your project.

Example reference in Kubebuilder source:
- https://github.com/kubernetes-sigs/kubebuilder/blob/f35565002e5378e14db20fd92d652f9de698f059/testdata/project-v4/cmd/main.go#L138-L144

Hope this helps!
